orttraining/orttraining/eager/ort_aten.cpp

```diff
@@ -682,6 +682,121 @@ at::Tensor& out) {
   return out;
 }
 
+/*
+ * Resize backing store of a TensorImpl.
+ *
+ * See notes for implementation details and potential differences from canonical implementations due to constraints in
+ * ORT model.
+ *
+ * If new size is the same size as existing tensor: reshape the existing tensor
+ * If new size is larger:  allocate new memory and copy over existing elements. New memory is uninitialized.
+ * If new size is smaller: allocate a smaller backing tensor, and copy over
+ *                         as many elements as will fit.
+ *
+ * Notes:
+ * There are some implementation details that might deviate from expectations:
+ *  - As the Onnxruntime::tensor does not support resize operation, this functionality is supported on the TensorImpl
+ *    by swapping out the backing tensor if the size changes.
+ *
+ *  - In the ORT model the shape of the TensorImpl is defined by the backing onnxruntime::tensor, so it is not supported
+ *    to have a TensorImpl with a different shape / size than the backing onnxruntime::tensor. This means when resizing
+ *    to a smaller TensorImpl, other implementations might keep the same backing storage, ORT will re-allocate a new
+ *    onnxruntime::tensor and copy over as many of the existing elements that fit. Functionally, you will end up with
+ *    same output, but the underlying buffer will be re-allocated.
+ *
+ *    A future change could be to allow ORTTensorImpl to have a different size / shape than the onnxrutime::tensor
+ *    backing it, and then we could improve this behavior.
+ *
+ * The canonical CPU / CUDA implementations in PyTorch repository:
+ *     CPU: aten/src/ATen/native/Resize.cpp
+ *     CUDA: aten/src/ATen/native/cuda/Resize.cpp
+ */
+void resize_impl_ort_(
+    onnxruntime::ORTInvoker& invoker,
+    ORTTensorImpl* self,
+    at::IntArrayRef size) {
+  auto self_ort_value = self->tensor();
+  auto old_shape = onnxruntime::TensorShape(self->sizes());
+  auto new_shape = onnxruntime::TensorShape(size.vec());
+
+  if (new_shape.Size() == old_shape.Size()) {
```

Comment:
what if the shape is already correct? Should we check for that and then do nothing?

Reply:
I think that would be reasonable. This condition is checked for in `resize_` - but seems reasonable that is should be done here as well, rather than expecting all callers to perform this check.